rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,473 @@
+
+*   Device registration: Will be addressed in a different RFC.
+*   Device management: The focus of this RFC.
+*   Kernel and op registration and implementation:
+    [RFC Accepted](https://github.com/tensorflow/community/blob/master/rfcs/20190814-kernel-and-op-registration.md). [C API implemented](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/).
+
+[StreamExecutor](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_pimpl.h;l=73) is TensorFlow's main device manager, responsible for work execution and memory management. It provides a set of methods
+(such as
+[Memcpy](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=240))
+that can be customized for a particular device.
+
+We propose a C API wrapper of a subset of methods in
+[StreamExecutorInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=166?q=StreamExecutorinterface)
+as an ABI-stable way to register a custom StreamExecutor platform.
+
+# User Benefits
+
+A decoupled way to add a new device to TensorFlow.
+
+*   Simpler process: Does not have to add a new build toolchain to TensorFlow
+*   Faster time-to-solution: Does not need code review from the TensorFlow team.
+*   Lower maintenance efforts: Only C-API-related changes could break the
+    integration. Unrelated TensorFlow changes would not break the code.
+       *    The C APIs may be changed during the initial experimental phase based 
+            on developer experience and feedback. When the APIs become more mature, 
+            we will try to keep them stable (in a best-effort manner) until the new 
+            TensorFlow stack is available.
+
+# Design Proposal
+
+## StreamExecutorInterface
+
+[StreamExecutorInterface](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/stream_executor_internal.h;l=166?q=StreamExecutorinterface)
+is quite large and some of its methods are only sporadically used. Therefore, we
+plan to wrap only a subset of key StreamExecutorInterface functionality.
+
+See proposed C API below:
+
+```cpp
+typedef SE_Stream_st* SE_Stream;
+typedef SE_Event_st* SE_Event;
+typedef SE_Timer_st* SE_Timer;
+typedef TF_Status* (*TF_StatusCallbackFn)(void*);
+
+typedef struct SE_PlatformId {
+ void* id;  // aka stream_executor::Platform::Id
+} SE_PlatformId;
+
+typedef struct SE_TimerFns {
+ const size_t struct_size;
+ uint64_t (*nanoseconds)(SE_Timer timer);
+ uint64_t (*microseconds)(SE_Timer timer);
+} SE_Timer;
+
+typedef struct SE_AllocatorStats {
+  const size_t struct_size;
+  int64_t num_allocs;
+  int64_t bytes_in_use;
+  int64_t peak_bytes_in_use;
+  int64_t largest_alloc_size;
+
+  bool has_bytes_limit;
+  int64_t bytes_limit;
+
+  int64_t bytes_reserved;
+  int64_t peak_bytes_reserved;
+
+  bool has_bytes_reservable_limit;
```

Comment:
`sizeof(bool)` is implementation defined https://stackoverflow.com/questions/4897844/is-sizeofbool-defined-in-the-c-language-standard

To guarantee binary stability you should probably use fixed-size type. 

That said, do we need the boolean flags here at all? Do values of 0 for `bytes_limit` and `bytes_reservable_limit` serve as `unset` ?

Reply:
Your suggestion to treat `0` as unset sounds reasonable. Looks like @tjablin tried to do that at some point: https://osscs.corp.google.com/tensorflow/tensorflow/+/9ca8321d68b74ce4f30756366919c80282ca2b68:tensorflow/stream_executor/allocator_stats.h;dlc=3c2033678feb046caede91832045ac8bacb2f95a which was rolled back.

I just asked him and he seems to think optional value is still a good idea for clarity. I guess we can use `int8_t` for has_bytes_* fields.

Reply:
Agree, please use int8_t or uint8_t instead of bool.

Reply:
+@yisitu submitted a change to define TF_BOOL as unsigned char.
We decided to use `unsigned char` for consistency with this comment: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/c_api.h#L42